### PR TITLE
fix: return error and reconnect network when the network lost

### PR DESCRIPTION
**Pull request by sulenn**

### CODE REVIEW

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

File: conn/channel.go [hunk 1]
```diff
@@ -417,6 +421,9 @@ func (hc *channelSession) doRPCRequest(ctx context.Context, msg interface{}) (io
 	}
 	msgBytes := rpcMsg.Encode()
 
+	if hc.c == nil {
+		return nil, errors.New("the current network is unavailable")
```
Comment (bxq2011hust) [inline conn/channel.go:425]:
```suggestion
		return nil, errors.New("connection lost")
```

File: conn/channel.go [hunk 2]
```diff
@@ -431,6 +438,9 @@ func (hc *channelSession) doRPCRequest(ctx context.Context, msg interface{}) (io
 	response = hc.responses[rpcMsg.uuid]
 	delete(hc.responses, rpcMsg.uuid)
 	hc.mu.Unlock()
+	if response.Err != nil { // network is not reliable
```
Comment (bxq2011hust) [inline conn/channel.go:441]:
```suggestion
	if response.Err != nil { // connection lost
```

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

File: conn/channel.go [hunk 3]
```diff
@@ -905,12 +933,46 @@ func (hc *channelSession) processMessages() {
...
+			hc.mu.Unlock()
+			// delete messageHandler
+			hc.messageMu.Lock()
+			for key := range hc.messageHandlers {
+				delete(hc.messageHandlers, key)
+			}
+			hc.messageMu.Unlock()
+			// delete old network
+			_ = hc.c.Close()
+			hc.c = nil
```
Comment (bxq2011hust) [inline conn/channel.go:956]: 先置空，在清理吧

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  6
         

Clones added
============
- conn/channel.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=6030440&bid=19763505)

**Review by bxq2011hust — APPROVED** [review]

**Pull request merged**